clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -908,311 +628,781 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
+    handleSpecialType(FD, ArgTy);
+  }
+
+  void handlePointerType(FieldDecl *FD, QualType ArgTy) final {
+    // TODO: Can we document what the heck this is doing?!
+    QualType PointeeTy = ArgTy->getPointeeType();
+    Qualifiers Quals = PointeeTy.getQualifiers();
+    Quals.setAddressSpace(LangAS::opencl_global);
+    PointeeTy = SemaRef.getASTContext().getQualifiedType(
+        PointeeTy.getUnqualifiedType(), Quals);
+    QualType ModTy = SemaRef.getASTContext().getPointerType(PointeeTy);
+    addParam(FD, ModTy);
+  }
+
+  void handleScalarType(FieldDecl *FD, QualType ArgTy) final {
+    addParam(FD, ArgTy);
+  }
+
+  // This is implemented here because this is the only case where the recurse
+  // object is required.  The base type is pretty cheap, so we might opt
+  // to just always create it (the way this one is implemented) and just put
+  // this implementation in the base.
+  void handleStructType(FieldDecl *FD, QualType ArgTy) final {
+    addParam(FD, ArgTy);
+  }
+
+  void handleSyclStreamType(FieldDecl *FD, QualType ArgTy) final {
+    addParam(FD, ArgTy);
+  }
+
+  void handleSyclStreamType(const CXXBaseSpecifier &, QualType ArgTy) final {
+    // TODO Implement
+  }
+
+  void setBody(CompoundStmt *KB) { KernelDecl->setBody(KB); }
+
+  FunctionDecl *getKernelDecl() { return KernelDecl; }
+
+  llvm::ArrayRef<ParmVarDecl *> getParamVarDeclsForCurrentField() {
+    return ArrayRef<ParmVarDecl *>(std::begin(Params) + LastParamIndex,
+                                   std::end(Params));
+  }
+};
+
+class SyclKernelBodyCreator
+    : public SyclKernelFieldHandler<SyclKernelBodyCreator> {
+  SyclKernelDeclCreator &DeclCreator;
+  llvm::SmallVector<Stmt *, 16> BodyStmts;
+  llvm::SmallVector<Stmt *, 16> FinalizeStmts;
+  llvm::SmallVector<Expr *, 16> InitExprs;
+  VarDecl *KernelObjClone;
+  InitializedEntity VarEntity;
+  CXXRecordDecl *KernelObj;
+  llvm::SmallVector<Expr *, 16> MemberExprBases;
+  FunctionDecl *KernelCallerFunc;
+
+  // Using the statements/init expressions that we've created, this generates
+  // the kernel body compound stmt. CompoundStmt needs to know its number of
+  // statements in advance to allocate it, so we cannot do this as we go along.
+  CompoundStmt *createKernelBody() {
+
+    Expr *ILE = new (SemaRef.getASTContext()) InitListExpr(
+        SemaRef.getASTContext(), SourceLocation(), InitExprs, SourceLocation());
+    ILE->setType(QualType(KernelObj->getTypeForDecl(), 0));
+    KernelObjClone->setInit(ILE);
+    Stmt *FunctionBody = KernelCallerFunc->getBody();
+
+    ParmVarDecl *KernelObjParam = *(KernelCallerFunc->param_begin());
+
+    // DeclRefExpr with valid source location but with decl which is not marked
+    // as used is invalid.
+    KernelObjClone->setIsUsed();
+    std::pair<DeclaratorDecl *, DeclaratorDecl *> MappingPair;
+    MappingPair.first = KernelObjParam;
+    MappingPair.second = KernelObjClone;
+
+    // Function scope might be empty, so we do push
```

Comment:
Is this comment incomplete?

Reply:
Why?

Reply:
It just doesn't end in a period and seems like a partial sentence.  Something like, "Push the Kernel function scope to ensure the scope isn't empty." might fit better.

Reply:
Okay.